src/Compilers/CSharp/Test/Semantic/Semantics/StackAllocInitializerTests.cs

```diff
@@ -97,25 +233,231 @@ public void Method1()
+            Assert.Equal(1, variables.Count());
+
+            var obj1 = variables.ElementAt(0);
+            Assert.Equal("obj1", obj1.Identifier.Text);
+
+            var obj1Value = model.GetSemanticInfoSummary(obj1.Initializer.Value);
+            Assert.Equal(SpecialType.System_Double, ((NamedTypeSymbol)obj1Value.Type).TypeArgumentsNoUseSiteDiagnostics[0].SpecialType);
+            Assert.Equal(SpecialType.System_Double, ((NamedTypeSymbol)obj1Value.ConvertedType).TypeArgumentsNoUseSiteDiagnostics[0].SpecialType);
+            Assert.Equal(ConversionKind.Identity, obj1Value.ImplicitConversion.Kind);
+            var declared = model.GetDeclaredSymbol(obj1.Initializer.Value);
+            Assert.Null(declared);
+        }
+
+        [Fact]
+        public void BadBestType_Pointer()
+        {
+            var comp = CreateCompilationWithMscorlibAndSpan(@"
+unsafe class Test
+{
+    ref struct S {}
+    void Method1(S s)
     {
         var obj1 = stackalloc[] { """" };
         var obj2 = stackalloc[] { new {} };
+        var obj3 = stackalloc[] { s }; // OK
     }
 }", TestOptions.UnsafeReleaseDll);
 
             comp.VerifyDiagnostics(
-                // (6,20): error CS0208: Cannot take the address of, get the size of, or declare a pointer to a managed type ('string')
+                // (7,20): error CS0208: Cannot take the address of, get the size of, or declare a pointer to a managed type ('string')
                 //         var obj1 = stackalloc[] { "" };
-                Diagnostic(ErrorCode.ERR_ManagedAddr, @"stackalloc[] { """" }").WithArguments("string").WithLocation(6, 20),
-                // (7,20): error CS0208: Cannot take the address of, get the size of, or declare a pointer to a managed type ('<empty anonymous type>')
+                Diagnostic(ErrorCode.ERR_ManagedAddr, @"stackalloc[] { """" }").WithArguments("string").WithLocation(7, 20),
+                // (8,20): error CS0208: Cannot take the address of, get the size of, or declare a pointer to a managed type ('<empty anonymous type>')
                 //         var obj2 = stackalloc[] { new {} };
-                Diagnostic(ErrorCode.ERR_ManagedAddr, "stackalloc[] { new {} }").WithArguments("<empty anonymous type>").WithLocation(7, 20)
+                Diagnostic(ErrorCode.ERR_ManagedAddr, "stackalloc[] { new {} }").WithArguments("<empty anonymous type>").WithLocation(8, 20)
+                );
+        }
+
+        [Fact]
+        public void BadBestType_Span()
+        {
+            var comp = CreateCompilationWithMscorlibAndSpan(@"
+unsafe class Test
+{
+    ref struct S {}
+    void Method1(S s, bool c)
+    {
+        var obj1 = c ? default : stackalloc[] { """" };
+        var obj2 = c ? default : stackalloc[] { new {} };
+        var obj3 = c ? default : stackalloc[] { s }; // Should be an error, see https://github.com/dotnet/roslyn/issues/25086
```

Comment:
This looks ok. When we merge the changes with the fix, the test will fails and we will update it.